<p style ="text-align:center">
    <img src="https://raw.githubusercontent.com/EduPekUfpr/PythonProject/refs/heads/main/Images/BannerUFPR.png" width="700" alt="PPGOLD/PPGMNE Python:INTRO"  />
</p>

## Prof. Eduardo Pécora, Ph.D.

# **Pandas: Valores Ausentes**

Neste notebook cobriremos os seguintes tópicos:

1. [Valores Ausentes](#oper)

<!-- NOTA PARA O INSTRUTOR: a imagem do banner agora aponta para o repositório do curso no GitHub (mesma fonte confiável usada para o CSV), em vez de um domínio pessoal que pode sair do ar. É necessário enviar o arquivo BannerUFPR.png para uma pasta 'Images/' nesse repositório para o link funcionar. -->

<a id="data"></a>
# **Leitura de Dados**

A leitura de dados é o primeiro passo em qualquer fluxo de trabalho de análise de dados. O Pandas facilita a importação de dados de várias fontes e formatos para um DataFrame, que é a estrutura de dados principal utilizada ao longo da análise.

**Arquivos CSV**

O formato CSV (Comma-Separated Values) é um dos formatos de dados mais comuns, utilizado amplamente para armazenar dados tabulares.

In [1]:
import pandas as pd

nba_file = "https://raw.githubusercontent.com/EduPekUfpr/PythonProject/refs/heads/main/Dados/nba.csv" #caminho do arquivo

df_nba = pd.read_csv(nba_file)
print(df_nba.head())  # Exibe as primeiras 5 linhas do DataFrame

            Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0  


**Primeira inspeção dos dados**

Antes de caçar valores ausentes, é uma boa prática olhar a estrutura geral do DataFrame: quantidade de linhas/colunas, tipos de dados e um resumo estatístico. Isso já dá pistas de onde podem estar os problemas (colunas com menos valores não-nulos que o total de linhas, por exemplo).

In [2]:
df_nba.info()  # tipos de dados e contagem de valores não-nulos por coluna
df_nba.describe()  # resumo estatístico das colunas numéricas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 457 entries, 0 to 456
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


,Number,Age,Weight,Salary
count,457.000000,457.000000,457.000000,4.460000e+02
mean,17.678337,26.938731,221.522976,4.842684e+06
std,15.966090,4.404016,26.368343,5.229238e+06
min,0.000000,19.000000,161.000000,3.088800e+04
25%,5.000000,24.000000,200.000000,1.044792e+06
50%,13.000000,26.000000,220.000000,2.839073e+06
75%,25.000000,30.000000,240.000000,6.500000e+06
max,99.000000,40.000000,307.000000,2.500000e+07


<a id="treat"></a>
# **Tratamento de Valores Ausentes e Duplicados**

Durante a análise de dados, é comum encontrar valores ausentes e dados duplicados. O tratamento adequado desses problemas é essencial para garantir a qualidade dos resultados da análise. O Pandas oferece diversas funções e métodos para identificar, lidar e substituir esses valores de maneira eficiente.

## Identificação de Valores Ausentes

Valores ausentes podem aparecer por vários motivos, como erros de coleta de dados ou ausência real de informação. No Pandas, valores ausentes são representados como `NaN` (Not a Number)

In [3]:
df_nba.isnull() #encontra do DF os valores ausentes. Pode-se usar o isna() também

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,True
3,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...
452,False,False,False,False,False,False,False,False,False
453,False,False,False,False,False,False,False,False,False
454,False,False,False,False,False,False,False,True,False
455,False,False,False,False,False,False,False,True,False


In [4]:
df_nba.isnull().sum() #Somando os booleanos podemos saber a quantidade de ausentes

,0
Name,0
Team,0
Number,0
Position,0
Age,0
Height,0
Weight,0
College,84
Salary,11


In [5]:
# Contagem absoluta é útil, mas o percentual ajuda a decidir a estratégia:
# poucos % ausentes -> geralmente dá para dropar; muitos % -> dropar destrói a amostra, melhor imputar
(df_nba.isnull().sum() / len(df_nba) * 100).round(2)

,0
Name,0.00
Team,0.00
Number,0.00
Position,0.00
Age,0.00
Height,0.00
Weight,0.00
College,18.38
Salary,2.41


## Remoção de Valores Ausentes

In [6]:
df_nba_dropRows = df_nba.dropna() #remove as linhas em que alguma coluna tem valor ausente
df_nba_dropRows.isnull().sum()

,0
Name,0
Team,0
Number,0
Position,0
Age,0
Height,0
Weight,0
College,0
Salary,0


In [7]:
df_nba_dropRows.shape

(364, 9)

In [8]:
df_nba_dropCols = df_nba.dropna(axis=1) #remove as colunas em que alguma linha tem valor ausente
df_nba_dropCols.isnull().sum()

,0
Name,0
Team,0
Number,0
Position,0
Age,0
Height,0
Weight,0


In [9]:
df_nba_dropCols.shape

(457, 7)

## Substituição de Valores Ausentes

⚠️ **Cuidado:** preencher *todo* o DataFrame com um único valor (como `0`) raramente é uma boa ideia. Colunas numéricas como `Salary` ganhariam um valor sem sentido (salário zero), e colunas de texto como `College` ganhariam um número onde deveria haver uma string. O exemplo abaixo serve só para mostrar **como o método funciona** — na prática, a escolha do valor de preenchimento deve ser feita **coluna por coluna**, como faremos logo em seguida.

In [10]:
#substitui valores nulos por algum valor estabelecido.
#Neste caso ZERO
df_nba.fillna(0)

,Name,Team,Number,Position,Age,Height,Weight,College,Salary
0,Avery Bradley,Boston Celtics,0.0,PG,25.0,6-2,180.0,Texas,7730337.0
1,Jae Crowder,Boston Celtics,99.0,SF,25.0,6-6,235.0,Marquette,6796117.0
2,John Holland,Boston Celtics,30.0,SG,27.0,6-5,205.0,Boston University,0.0
3,R.J. Hunter,Boston Celtics,28.0,SG,22.0,6-5,185.0,Georgia State,1148640.0
4,Jonas Jerebko,Boston Celtics,8.0,PF,29.0,6-10,231.0,0,5000000.0
...,...,...,...,...,...,...,...,...,...
452,Trey Lyles,Utah Jazz,41.0,PF,20.0,6-10,234.0,Kentucky,2239800.0
453,Shelvin Mack,Utah Jazz,8.0,PG,26.0,6-3,203.0,Butler,2433333.0
454,Raul Neto,Utah Jazz,25.0,PG,24.0,6-1,179.0,0,900000.0
455,Tibor Pleiss,Utah Jazz,21.0,C,26.0,7-3,256.0,0,2900000.0


In [11]:
df_nba["Salary"].fillna(df_nba["Salary"].mean())

,Salary
0,7.730337e+06
1,6.796117e+06
2,4.842684e+06
3,1.148640e+06
4,5.000000e+06
...,...
452,2.239800e+06
453,2.433333e+06
454,9.000000e+05
455,2.900000e+06


In [12]:
df_nba["College"].fillna("UFPR")

,College
0,Texas
1,Marquette
2,Boston University
3,Georgia State
4,UFPR
...,...
452,Kentucky
453,Butler
454,UFPR
455,UFPR


In [13]:
df_nba.isnull().sum()

,0
Name,0
Team,0
Number,0
Position,0
Age,0
Height,0
Weight,0
College,84
Salary,11


**Por que dropar `Salary` mas imputar `College`?**

A decisão não é arbitrária — depende do papel de cada coluna na análise:

- `Salary` é uma variável **numérica central** para qualquer análise financeira do dataset. Inventar um valor para ela (média, zero, etc.) distorceria estatísticas e modelos que a usem como alvo ou feature. Como a proporção de ausentes é pequena (veja o percentual calculado acima), **dropar as linhas** é a opção mais segura.
- `College` é uma variável **categórica** usada mais para agrupamento/contexto do que para cálculo numérico. Preencher com um rótulo distinto (`'UFPR'`) preserva a linha (e o salário, altura, etc. do jogador) sem fingir que sabemos qual foi a faculdade — só marcamos explicitamente que essa informação não estava disponível.

Ou seja: **a estratégia depende de quão crítica a coluna é para a análise e de quanto dado se perderia ao dropar.**

In [14]:
# melhor opção neste caso,
df_nba = df_nba.dropna(subset=['Salary']) # 'dropar' as linhas com valores ausentes na coluna salário
df_nba["College"] = df_nba["College"].fillna("UFPR") # Substituir por um valor diferente dos demais na coluna 'college'
df_nba.isnull().sum()
df_nba.shape

(446, 9)

In [15]:
df_nba.isnull().sum()

,0
Name,0
Team,0
Number,0
Position,0
Age,0
Height,0
Weight,0
College,0
Salary,0


<a id="dup"></a>
## Identificação e Remoção de Duplicados

Além de valores ausentes, é comum encontrar **linhas duplicadas** em um dataset — registros repetidos por erro de coleta, junções (`merge`) mal feitas, ou reenvio do mesmo dado. O Pandas oferece `duplicated()` para identificar e `drop_duplicates()` para remover.

In [16]:
df_nba.duplicated().sum()  # quantidade de linhas totalmente duplicadas

np.int64(0)

Se houver duplicados, vale inspecionar quais linhas são antes de decidir removê-las — às vezes uma aparente duplicata é, na verdade, um dado legítimo (por exemplo, dois jogadores diferentes com o mesmo nome). Por padrão, `duplicated()` marca como `True` a partir da **segunda** ocorrência em diante.

In [17]:
df_nba[df_nba.duplicated(keep=False)]  # mostra TODAS as ocorrências envolvidas em duplicação (não só a repetida)

,Name,Team,Number,Position,Age,Height,Weight,College,Salary


Também é possível considerar duplicado com base em **algumas colunas específicas**, não a linha inteira — útil quando, por exemplo, duas linhas têm o mesmo jogador (`Name`) mas com pequenas diferenças em outras colunas por erro de digitação.

In [18]:
# Exemplo: considerar duplicado apenas pelo nome do jogador
df_nba.duplicated(subset=['Name']).sum()

np.int64(0)

In [19]:
# Remove duplicados, mantendo a primeira ocorrência (padrão: keep='first')
df_nba = df_nba.drop_duplicates()
df_nba.shape

(446, 9)

## Fique Conectado

- [![YouTube](https://img.icons8.com/?size=40&id=19318&format=png&color=000000)](https://www.youtube.com/@LigaDataScience/videos)  
  Explore nossos vídeos educacionais e webinars sobre ciência de dados, machine learning e inteligência artificial. Inscreva-se para não perder nenhuma atualização!

- [![LinkedIn](https://img.icons8.com/?size=40&id=13930&format=png&color=000000)](https://www.linkedin.com/company/liga-data-science-ufpr/)  
  Siga-nos no LinkedIn para as últimas novidades, oportunidades de carreira e networking profissional no campo da ciência de dados.

- [![Instagram](https://img.icons8.com/?size=40&id=32323&format=png&color=000000)](https://www.instagram.com/ligadatascience/)  
  Confira nosso Instagram para conteúdos dos bastidores, destaques de eventos e o dia a dia da Liga Data Science. Faça parte da nossa jornada!
  
  ## Autores

<a href="https://www.linkedin.com/in/eduardopecora/" target="_blank">Eduardo Pecora</a>

<a href="https://www.linkedin.com/in/jo%C3%A3o-gabriel-santin-botelho-618244222/" target="_blank">João Gabriel Santin Botelho</a>

## Log de modificações

| Data | Versão | Modificado por | Descrição |
| ----------------- | ------- | ---------- | ---------------------------------- |
| 29-08-2024       | 1.0     | Eduardo Pecora & João Gabriel| Inicial               |
| 10-08-2026       | 1.1     | Eduardo Pecora | Revisão               |

<hr>

## <h3 align="center"> (c) Liga Data Science/ UFPR 2024. All rights reserved. <h3/>